From 506500ee2dfdfdc3c5f076847352d337f3795dbb Mon Sep 17 00:00:00 2001
From: Pzqqt <821026875@qq.com>
Date: Thu, 23 May 2024 20:12:51 +0800
Subject: [PATCH] crypto: zstd: support negative compression level

zstd supports negative compression level, which will further reduce
the compression ratio and increase the compression and decompression
speed.

But please note that according to [zstd official benchmark results]
(https://github.com/facebook/zstd#benchmarks), when you set the
compression level to "-3", the compression ratio and compression
speed are very close to lz4, and the decompression speed is only
about 50% of lz4.

Therefore, it is best to use lz4 at this time, and it also tells us
not to blindly set too low a compression level for zstd.
---
 crypto/zstd.c | 9 ++++-----
 1 file changed, 4 insertions(+), 5 deletions(-)

diff --git a/crypto/zstd.c b/crypto/zstd.c
index cea47194a9e89..1b248a1ccf263 100644
--- a/crypto/zstd.c
+++ b/crypto/zstd.c
@@ -15,18 +15,17 @@
 #include <crypto/internal/scompress.h>
 
 
-static uint __read_mostly compression_level = 3;
+static int __read_mostly compression_level = 3;
 
 static int set_compression_level(const char *buf, const struct kernel_param *kp)
 {
-	int ret;
-	uint temp;
+	int ret, temp;
 
-	ret = kstrtouint(buf, 0, &temp);
+	ret = kstrtoint(buf, 0, &temp);
 	if (ret)
 		return ret;
 
-	if (temp == 0 || temp > zstd_max_clevel())
+	if (temp == 0 || temp < zstd_min_clevel() || temp > zstd_max_clevel())
 		return -EINVAL;
 
 	return param_set_int(buf, kp);
